src/history.cpp

```diff
@@ -0,0 +1,49 @@
+/*
+ * history.cpp
+ *
+ *  Created on: 08 Apr 2021
+ *      Author: gc1610
+ */
+
+#include <Rcpp.h>
+#include <algorithm>
+#include "history.h"
+
+History::History() : max_size(-1) {}
+History::History(size_t max_size) : max_size(max_size) {}
+
+void History::push(double time, double value) {
+    values.insert({time, value});
+    if (max_size != -1) {
+        while(values.size() > max_size) {
+            values.erase(values.begin());
+        }
+    }
+}
+
+double History::at(double time) const {
+    auto it = values.lower_bound(time);
+    if (it == values.end()) {
+        Rcpp::stop("`time` is later than the stored history");
+    }
+
+    // Check if we've landed on the exact time
+    if (it->first == time) {
+        return it->second;
+    }
+
+    // Find the element before
+    auto after_element = *it;
+    while(it->first > time) {
+        // Check if we're at the start of the history
+        if (it == values.begin()) {
+            Rcpp::stop("`time` is before our stored history");
+        }
+        it--;
+    }
+
+    // Interpolate
+    return it->second + (
```

Comment:
It looks like here you are doing a linear interpolation of a variable. Depending on the dynamics of the system this may cause your error calculation to lock up because the linear error here (which will echo through the solution) will be impossible for it to eliminate even as it reduces the step size, as the error is already baked in. I am not sure in practice how much of a problem this is, especially if the mosquito system is simple - it's a bit like having a forcing function that is piecewise linear, which we include all the time.

This is the interface for "dense output" which allows interpolation within a timestep with the odint solvers:
https://www.boost.org/doc/libs/1_60_0/libs/numeric/odeint/doc/html/boost_numeric_odeint/concepts/dense_output_stepper.html

However, it's not at all clear to me how you'd be able to push back the that state each step.

Reply:
I think this is important, so I'm gonna delay 😉 this fix until I get some time to figure out the most efficient way to do this.

My hope is that I can keep a history in boost's stepper and use `calc_state`. At the moment, I don't think the boost's stepper has the history.